rbx_types/src/attributes/writer.rs

```diff
@@ -0,0 +1,136 @@
+use super::*;
+
+use crate::{basic_types::*, variant::Variant};
+
+use std::{borrow::Borrow, io::Write};
+
+fn write_f32<W: Write>(mut writer: W, n: f32) {
+    writer
+        .write_all(&n.to_le_bytes().to_vec())
+        .expect("couldn't write f32 to buffer");
+}
+
+fn write_u32<W: Write>(mut writer: W, n: u32) {
+    writer
+        .write_all(&n.to_le_bytes().to_vec())
+        .expect("couldn't write u32 to buffer");
+}
+
+fn write_color3<W: Write>(mut writer: W, color: Color3) {
+    write_f32(&mut writer, color.r);
+    write_f32(&mut writer, color.g);
+    write_f32(&mut writer, color.b);
+}
+
+fn write_string<T: AsRef<[u8]>, W: Write>(mut writer: W, string: T) {
+    let bytes = string.as_ref();
+    write_u32(&mut writer, bytes.len() as u32);
+    writer
+        .write_all(bytes)
+        .expect("couldn't write string to buffer");
+}
+
+fn write_udim<W: Write>(mut writer: W, udim: UDim) {
+    write_f32(&mut writer, udim.scale);
+    writer
+        .write_all(&udim.offset.to_le_bytes().to_vec())
+        .expect("couldn't write UDim to buffer");
+}
+
+fn write_vector2<W: Write>(mut writer: W, vector2: Vector2) {
+    write_f32(&mut writer, vector2.x);
+    write_f32(&mut writer, vector2.y);
+}
+
+/// Writes the attribute property (AttributesSerialize) from a map of attribute names -> values.
+pub(crate) fn attributes_from_map<
+    K: Borrow<String>,
+    V: Borrow<Variant>,
+    I: Iterator<Item = (K, V)> + ExactSizeIterator,
+    M: IntoIterator<IntoIter = I, Item = (K, V)>,
+>(
+    map: M,
+) -> Result<Vec<u8>, AttributeError> {
+    let map = map.into_iter();
+    let mut bytes = Vec::new();
+
+    bytes
+        .write_all(&(map.len() as u32).to_le_bytes())
+        .expect("couldn't write map length to buffer");
+
+    for (name, variant) in map {
+        let variant = variant.borrow();
+
+        write_string(&mut bytes, name.borrow());
+
+        let attribute_type = AttributeType::try_from(variant.ty())?;
```

Comment:
Why do we need `AttributeType` here? Why not just match on `Variant` and then return an error for variants that we don't match?

Reply:
It's to then push its type as a byte.